modules/ximgproc/src/globalmatting.cpp

```diff
@@ -0,0 +1,568 @@
+{
+    float result = 0;
+    float div = 1e-6f;
+    for (int c = 0; c < 3; ++c)
+    {
+	float f = F[c];
+	float b = B[c];
+	float i = I[c];
+
+	result += (i - b) * (f - b);
+	div += (f - b) * (f - b);
+    }
+
+    return std::min(std::max(result / div, 0.f), 1.f);
+}
+
+// Eq. 3
+float GlobalMatting::colorCost(const cv::Vec3b &F, const cv::Vec3b &B, const cv::Vec3b &I, float alpha)
+{
+    float result = 0;
+    for (int c = 0; c < 3; ++c)
+    {
+	float f = F[c];
+	float b = B[c];
+	float i = I[c];
+
+	result += sqr(i - (alpha * f + (1 - alpha) * b));
+    }
+
+    return sqrt(result);
+}
+
+// Eq. 4
+float GlobalMatting::distCost(const cv::Point &p0, const cv::Point &p1, float minDist)
+{
+    int dist = sqr(p0.x - p1.x) + sqr(p0.y - p1.y);
+    return sqrt((float)dist) / minDist;
+}
+
+float GlobalMatting::colorDist(const cv::Vec3b &I0, const cv::Vec3b &I1)
+{
+    int result = 0;
+
+    for (int c = 0; c < 3; ++c)
+	result += sqr((int)I0[c] - (int)I1[c]);
+
+    return sqrt((float)result);
+}
+
+float GlobalMatting::nearestDistance(const std::vector<cv::Point> &boundary, const cv::Point &p)
+{
+    int minDist2 = INT_MAX;
+    for (std::size_t i = 0; i < boundary.size(); ++i)
+    {
+	int dist2 = sqr(boundary[i].x - p.x)  + sqr(boundary[i].y - p.y);
+	minDist2 = std::min(minDist2, dist2);
+    }
+
+    return sqrt((float)minDist2);
+}
+
+
+
+
+void GlobalMatting::expansionOfKnownRegions(const cv::Mat_<cv::Vec3b> &image,
+	                            cv::Mat_<uchar> &trimap,
+	                            int r, float c)
+{
+    int w = image.cols;
+    int h = image.rows;
+
+    for (int x = 0; x < w; ++x)
+	for (int y = 0; y < h; ++y)
+	{
+	    if (trimap(y, x) != 128)
+	        continue;
+
+	    const cv::Vec3b &I = image(y, x);
+
+	    for (int j = y-r; j <= y+r; ++j)
+	        for (int i = x-r; i <= x+r; ++i)
+	        {
+	            if (i < 0 || i >= w || j < 0 || j >= h)
+	                continue;
+
+	            if (trimap(j, i) != 0 && trimap(j, i) != 255)
+	                continue;
+
+	            const cv::Vec3b &I2 = image(j, i);
+
+	            float pd = sqrt((float)(sqr(x - i) + sqr(y - j)));
+	            float cd = colorDist(I, I2);
+
+	            if (pd <= r && cd <= c)
+	            {
+	                if (trimap(j, i) == 0)
+	                    trimap(y, x) = 1;
+	                else if (trimap(j, i) == 255)
+	                    trimap(y, x) = 254;
+	            }
+	        }
+	}
+
+    for (int x = 0; x < trimap.cols; ++x)
+	for (int y = 0; y < trimap.rows; ++y)
+	{
+	    if (trimap(y, x) == 1)
+	        trimap(y, x) = 0;
+	    else if (trimap(y, x) == 254)
+	        trimap(y, x) = 255;
+
+	}
+}
+
+// erode foreground and background regions to increase the size of unknown region
+void GlobalMatting::erodeFB(cv::Mat_<uchar> &trimap, int r)
+{
+    int w = trimap.cols;
+    int h = trimap.rows;
+
+    cv::Mat_<uchar> foreground(trimap.size(), (uchar)0);
+    cv::Mat_<uchar> background(trimap.size(), (uchar)0);
```

Comment:
> (uchar)0

`Scalar::all(0)`

Reply:
Scalar::all(0) works on my local machine but is producing errors here. @alalek

Reply:
Right, it is template `Mat_` - my mistake. I though it is workaround for 0 implicit cast to `void*`.